File: 2025/JuyoungMoon/week5/2025-07-06-linked-list-cycle-ii.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 5: Linked List
// https://github.com/juyomo/leetcode-study

// LeetCode #142.
// https://leetcode.com/problems/linked-list-cycle-ii/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return nullptr;
        }
        
        unordered_set<ListNode*> seen;
        
        ListNode* curr = head;
        while (curr != nullptr) {
            if (seen.find(curr) != seen.end()) {
                return curr;
            }
            seen.insert(curr);
            curr = curr->next;
        }
        return nullptr;
    }
};

File: 2025/shinyoung/2025-07-06-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow is fast:
                ptr1 = head
                ptr2 = slow
                while ptr1 is not ptr2:
                    ptr1 = ptr1.next
                    ptr2 = ptr2.next
                return ptr1
        return None
